lib/Shell.js

```diff
@@ -0,0 +1,239 @@
+const { EventEmitter } = require('events');
+const { spawn } = require('child_process');
+const os = require('os');
+const shortid = require('shortid');
+
+const PSCommand = require('./PSCommand');
+
+const {
+  isWin, isCriticalPSError, logger,
+  ShellStreamBuffer, shellSafeWrite,
+  getType, convertToPSOption,
+} = require('./utils');
+
+const {
+  PS_PROC_ERROR,
+  PS_ARG_MISS_ERROR,
+  PS_ARG_TYPE_ERROR,
+  PS_CMD_FAIL_ERROR,
+} = require('./errors');
+
+class Shell extends EventEmitter {
+  constructor(options = {}) {
+    super();
+
+    this.pid = -1;
+    this.streams = [];
+    this.commands = [];
+    this.history = [];
+    this.hadErrors = false;
+    this.invocationStateInfo = 'NotStarted';
+    this.verbose = false;
+
+    let psProc = !isWin ? 'pwsh' : 'powershell';
+    let psOpts = ['-NoLogo', '-NoExit', '-Command', '-'];
+
+    if(options.pwsh) {
+      psProc = 'pwsh';
+    }
+
+    if(options.pwshPrev && !options.pwsh) {
+      psProc = 'pwsh-preview';
+    }
+
+    if(process.env.NPS) {
+      psProc = process.env.NPS;
+    }
+
+    if(Object.hasOwnProperty.call(options, 'debugMsg')) {
+      logger.warn(`debugMsg will be deprecated soon. use verbose instead.\n
+      https://rannn505.gitbook.io/node-powershell/shell/initialize`);
+      this.verbose = options.debugMsg;
+    }
+
+    if(options.verbose) {
+      this.verbose = true;
+    }
+
+    Object.keys(options).forEach(optKey => {
```

Comment:
Not sure why the explicit loop here vs something like:

```js
const speciallyHandled = ['pwsh', 'pwshPrev', 'debugMsg', 'verbose', 'inputEncoding', 'outputEncoding'];
const options = Object.entries(options)
    .filter(([k,v]) => !speciallyHandled.includes(k))
   .map(([k, v]) => convertToPSOption(k, v));
psOpts = psOpts.concat(options);
```

Reply:
:1st_place_medal: